server/project_member.go

```diff
@@ -2,18 +2,161 @@ package server
+		}
+		roleProvider := api.ProjectRoleProvider(c.Param("roleProvider"))
+
+		switch roleProvider {
+		case api.ProjectRoleProviderGitlabSelfHost:
+			{
+				projectFind := &api.ProjectFind{ID: &projectID}
+				project, err := s.ProjectService.FindProject(ctx, projectFind)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Project not found: %s", c.Param("projectID"))).SetInternal(err)
+				}
+				if project.WorkflowType == api.UIWorkflow {
+					return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid workflow type: %s", project.WorkflowType))
+				}
+
+				repoFind := &api.RepositoryFind{ProjectID: &projectID}
+				repo, err := s.RepositoryService.FindRepository(ctx, repoFind)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Fail to fetch relevant VCS repo, Project ID: %s", c.Param("projectID"))).SetInternal(err)
+				}
+
+				vcsFind := &api.VCSFind{
+					ID: &repo.VCSID,
+				}
+				vcs, err := s.VCSService.FindVCS(ctx, vcsFind)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find VCS for sync project member: %d", repo.VCSID)).SetInternal(err)
+				}
+				if vcs == nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("VCS ID not found: %d", repo.VCSID))
+				}
+
+				gitlabProjectMemberList, err := vcsPlugin.Get("GITLAB_SELF_HOST", vcsPlugin.ProviderConfig{Logger: s.l}).FetchProjectMemberList(ctx,
+					common.OauthContext{
+						ClientID:     vcs.ApplicationID,
+						ClientSecret: vcs.Secret,
+						AccessToken:  repo.AccessToken,
+						RefreshToken: repo.RefreshToken,
+						Refresher:    s.refreshToken(ctx, repo.ID),
+					},
+					vcs.InstanceURL,
+					repo.ExternalID,
+				)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("fail to fetch project member from GitLab, instance URL: %s", vcs.InstanceURL)).SetInternal(err)
+				}
+
+				// create or patch project member
+				findProjectMember := &api.ProjectMemberFind{ProjectID: &projectID}
+				bytebaseProjectMemberList, err := s.ProjectMemberService.FindProjectMemberList(ctx, findProjectMember)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("fail to fetch project member in bytebase: %d", projectID)).SetInternal(err)
+				}
+
+				// check whether principal exists in our system. if not exist, create one.
+				for _, projectMember := range gitlabProjectMemberList {
+					findPrincipal := &api.PrincipalFind{Email: &projectMember.Email}
+					principal, err := s.PrincipalService.FindPrincipal(ctx, findPrincipal)
+					if err != nil {
+						return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal info").SetInternal(err)
+					}
+
+					if projectMember.State != vcsPlugin.UserStateActive && projectMember.MembershipState != vcsPlugin.UserStateActive {
```

Comment:
Seems like this should be handled inside the vcs.Fetch method, the vcs.Fetch would only return the active user.